Allow injecting a profiler runtime into `#![no_core]` crates

Zalathar opened the pull request:
An alternative to #133300, allowing `-Cinstrument-coverage` to be used with `-Zbuild-std`.

The incompatibility between `profiler_builtins` and `#![no_core]` crates appears to have been caused by profiler_builtins depending on core, and therefore conflicting with core (or minicore).

But that's a false dependency, because the profiler doesn't contain any actual Rust code. So we can just mark the profiler itself as `#![no_core]`, and remove the incompatibility error.

---

For context, the error was originally added by #79958.

Zalathar committed:
Sort and separate lint/feature attributes in `profiler_builtins`

Zalathar committed:
Remove unnecessary `#![allow(unused_features)]`

Zalathar committed:
Make profiler_builtins `#![no_core]` instead of just `#![no_std]`

This crate doesn't contain any actual Rust code; it's just C/C++ code built and
packaged in a Rust-friendly way.

Zalathar committed:
Allow injecting a profiler runtime into `#![no_core]` crates

Now that the profiler runtime is itself `#![no_core]`, it can be a dependency
of other no_core crates, including core.

Zalathar committed:
Clarify logic for whether a profiler runtime is needed

Zalathar commented:
cc @Enselic @jieyouxu

Zalathar committed:
Coverage test for allowing coverage in a `#![no_core]` crate

jieyouxu commented:
I can't say I'm too familiar with `profiler_builtins` injection either, but cutting off the unnecessary dependency to `core` seems like a less fragile option that only skipping when building an `.rlib`.

bjorn3 commented on library/profiler_builtins/src/lib.rs:
```diff
@@ -1,11 +1,15 @@
+// tidy-alphabetical-end
+
+// Other attributes:
+#![no_core]
```
You can also remove the core and compiler_builtins dependencies in Cargo.toml.

Zalathar commented:
r? compiler

It's possible to write a run-make test that actually runs cargo with `-Zbuild-std` and `RUSTFLAGS="-Cinstrument-coverage"`, but it's a bit fiddly and takes 20+ seconds to run (since it builds its own libraries), so I'm not sure whether it's worth adding.

jieyouxu commented:
r? jieyouxu

jieyouxu commented on compiler/rustc_metadata/src/creader.rs:
```diff
@@ -799,7 +799,7 @@ impl<'a, 'tcx> CrateLoader<'a, 'tcx> {
     }
 
-    fn inject_profiler_runtime(&mut self, krate: &ast::Crate) {
+    fn inject_profiler_runtime(&mut self, _krate: &ast::Crate) {
```
Question: if `_krate` isn't used... maybe we should remove it?

Zalathar replied:
Yeah, in hindsight my reasons for leaving it as-is don't make much sense, so I'll change it.

jieyouxu reviewed (approved):
Thanks, this seems reasonable to me, just one nit. You can r=me with the nit.

jieyouxu commented:
@rustbot author

jieyouxu commented:
@bors rollup=never (profile runtime setup modifications)

Zalathar commented:
Addressed review feedback [(diff)](https://github.com/rust-lang/rust/compare/bd7aa4bcda901b7654b1552b04df667aa6a04387..6798ecaf1000c223378151023b8eb37f3464af29).

@bors r=jieyouxu

bors commented:
:pushpin: Commit 6798ecaf1000c223378151023b8eb37f3464af29 has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu 6798ecaf1000c223378151023b8eb37f3464af29 -->
<!-- homu: {"type":"Approved","sha":"6798ecaf1000c223378151023b8eb37f3464af29","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 6798ecaf1000c223378151023b8eb37f3464af29 with merge c1ad3840bb882f8d5030bab41c9d60a2de0ccb96...
<!-- homu: {"type":"BuildStarted","head_sha":"6798ecaf1000c223378151023b8eb37f3464af29","merge_sha":"c1ad3840bb882f8d5030bab41c9d60a2de0ccb96"} -->

rust-log-analyzer commented:
The job **`x86_64-msvc-ext`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/11997242409/job/33442646492) [(plain)](https://github.com/rust-lang-ci/rust/commit/c1ad3840bb882f8d5030bab41c9d60a2de0ccb96/checks/33442646492/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11997242409/job/33443754093)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/11997242409/job/33443754093","builder_name":"checks-actions"} -->

jieyouxu commented:
@bors retry (failed to pay msvc taxes)

bors commented:
:hourglass: Testing commit 6798ecaf1000c223378151023b8eb37f3464af29 with merge 39cb3386ddc6c71657418be28dbb3987eea4aa4b...
<!-- homu: {"type":"BuildStarted","head_sha":"6798ecaf1000c223378151023b8eb37f3464af29","merge_sha":"39cb3386ddc6c71657418be28dbb3987eea4aa4b"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12048244101/job/33600484663)
Approved by: jieyouxu
Pushing 39cb3386ddc6c71657418be28dbb3987eea4aa4b to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jieyouxu","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12048244101/job/33600484663"},"merge_sha":"39cb3386ddc6c71657418be28dbb3987eea4aa4b"} -->

rust-timer commented:
Finished benchmarking commit (39cb3386ddc6c71657418be28dbb3987eea4aa4b): [comparison URL](https://perf.rust-lang.org/compare.html?start=5f8a2405a6a7ea0ff85072b3bf90f4cff1144e85&end=39cb3386ddc6c71657418be28dbb3987eea4aa4b&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=5f8a2405a6a7ea0ff85072b3bf90f4cff1144e85&end=39cb3386ddc6c71657418be28dbb3987eea4aa4b&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 1.2% | [1.2%, 1.2%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=5f8a2405a6a7ea0ff85072b3bf90f4cff1144e85&end=39cb3386ddc6c71657418be28dbb3987eea4aa4b&stat=max-rss)
<details>
<summary>Results (primary 3.6%, secondary -0.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.6%  | [3.6%, 3.6%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 1.1%  | [1.1%, 1.1%]   | 2     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.2% | [-2.2%, -2.2%] | 1     |
| All ❌✅ (primary)                 | 3.6%  | [3.6%, 3.6%]   | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=5f8a2405a6a7ea0ff85072b3bf90f4cff1144e85&end=39cb3386ddc6c71657418be28dbb3987eea4aa4b&stat=cycles:u)
<details>
<summary>Results (primary 4.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 4.2% | [4.2%, 4.2%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 4.2% | [4.2%, 4.2%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=5f8a2405a6a7ea0ff85072b3bf90f4cff1144e85&end=39cb3386ddc6c71657418be28dbb3987eea4aa4b&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 794.092s -> 793.948s (-0.02%)
**Artifact size**: 336.11 MiB -> 336.16 MiB (0.01%)
